### PR TITLE
fix(reporting): stop corrupting non-ASCII characters in streamed data

### LINKED ISSUE #180: Report data corrupts non-ASCII characters: chunked UTF-8 decode in handleInputStreamToOutputStream
## Summary

Report calculation results corrupt non-ASCII text. The same value can appear intact in one row and mangled in another within a single result set, because the corruption depends on the byte offset of the character in the stream rather than on the character itself.

The cause is in our own code, not in SQS/SQLite: `handleInputStreamToOutputStream` decodes each read chunk as if it were a self-contained UTF-8 document.

## Symptom

A report column holding Devanagari text renders as e.g. `१ तासापेक्ष<?><?><?> जास्त` instead of `१ तासापेक्षा जास्त` — one character replaced by exactly three U+FFFD replacement characters. Two rows carrying the identical enum value can show as two different values.

## Root cause

`application/aam-backend-service/src/main/kotlin/com/aamdigital/aambackendservice/common/stream/StreamHelper.kt:20`

```kotlin
val buffer = ByteArray(byteArrayBufferLength)
while ((inputStream.read(buffer).also { bytesRead = it }) != -1) {
    if (bytesRead > 0) {
        writer.write(String(buffer, 0, bytesRead, Charsets.UTF_8))
    }
}
```

Each chunk is decoded in isolation and then re-encoded. When a multi-byte UTF-8 sequence straddles a read boundary, the trailing partial bytes of chunk N and the leading continuation bytes of chunk N+1 are each decoded as invalid input, and the resulting `EF BF BD` bytes are written to the output permanently.

This explains the 3-for-1 signature exactly: `ा` (U+093E) is `E0 A4 BE`. Split after `E0`, chunk N yields one U+FFFD (incomplete sequence) and chunk N+1 yields two more (`A4` and `BE` are not valid start bytes) — one character in, three replacement characters out.

That the damage is *position*-dependent rather than *value*-dependent is what rules out a storage or collation encoding problem upstream: that would corrupt every instance of a value identically.

## Frequency

This is not "once per 4096 bytes". `InputStream.read(buffer)` may return fewer bytes than requested, and both call sites' sources do so routinely — the SQS response `Resource.inputStream` short-reads at HTTP/TCP chunk boundaries, and `SequenceInputStream.read()` never reads across sub-stream boundaries. Every short read is a corruption opportunity.

## Impact

The helper sits on the **write** path as well as the read path:

- `common/couchdb/core/CouchDbFileStorage.kt:45` — pipes the SQS response through it into the `data.json` CouchDB attachment. **Stored attachments are already corrupted on disk.**
- `reporting/reportcalculation/controller/ReportCalculationController.kt:193` and `:262` — corrupt a second time on read, at different offsets, because the prepended JSON envelope shifts the boundaries.

Consequences:

- Any report result containing non-ASCII text (any non-Latin script, but also accented Latin, curly quotes, emoji) may be silently corrupted.
- `dataHash` / attachment digests are computed over already-corrupted bytes, so the corruption is invisible to integrity checks.
- Downstream consumers of the reporting API receive the corrupted values.

The helper is generic infrastructure, so template export and any other future user of `FileStorage` are affected equally.

## Reproduction

Feed the helper an input where a multi-byte character straddles the buffer boundary and compare output bytes to input bytes:

```
input 4201 bytes -> output 4207 bytes, byte-identical: false
```

The 6 extra bytes are one 3-byte character replaced by three 3-byte U+FFFD sequences.

## Fix

The decode/re-encode round trip serves no purpose — the helper is a byte pipe and should copy bytes verbatim:

```kotlin
inputStream.copyTo(outputStream, byteArrayBufferLength)
```

There is currently no test covering this helper; a regression test asserting byte-for-byte equality across a boundary-straddling multi-byte character should be added with the fix.

## Remediation for existing data

Because the corruption is applied on write, deploying the fix does not repair stored results. Existing report calculations must be re-run after the fix is deployed. Affected attachments can be identified by fetching `data.json` directly from CouchDB (bypassing the API) and searching for `EF BF BD`.

Note when planning that recalculation: repairing an attachment changes its digest, and `ReportCalculationChangeUseCase` notifies webhook subscribers whenever the digest changes. Re-running the affected reports will therefore fire webhook notifications to downstream consumers for every report that contained non-ASCII text.

### PR DESCRIPTION
**Pull request by sleidig**
Fixes #180

## Problem

`handleInputStreamToOutputStream` decoded each read chunk as if it were a self-contained UTF-8 document and re-encoded it:

```kotlin
writer.write(String(buffer, 0, bytesRead, Charsets.UTF_8))
```

A multi-byte character straddling a read boundary is therefore split across two chunks, and each half decodes to replacement characters — permanently, because what gets written out is the `EF BF BD` encoding rather than the original bytes.

Because the damage depends on where a character lands relative to a read boundary rather than on the character itself, identical values could render differently within a single result set. Devanagari made this obvious: `ा` (U+093E, `E0 A4 BE`) split after its first byte yields one replacement character at the end of one chunk and two at the start of the next — one character in, three out.

This is not limited to large payloads. `read(buffer)` returns short far more often than the buffer size suggests: HTTP response streams split at network chunk boundaries and `SequenceInputStream` never reads across its sub-streams.

## Change

The payloads relayed through this helper are already encoded (UTF-8 JSON from SQS and CouchDb), so they are now copied as bytes. The signature and default buffer length are unchanged, so no call site moves.

**One deliberate behaviour change worth reviewing:** the old implementation closed the output stream through `OutputStreamWriter(...).use { }`; the new one only flushes. Both call sites own their stream's lifecycle and must not have it closed by the callback — Spring's `StreamingHttpOutputMessage.Body` callback in `CouchDbFileStorage`, and `StreamingResponseBody` in `ReportCalculationController`. The e2e suite writing and then reading back a real CouchDb attachment passes, which exercises both.

## Tests

Added `StreamHelperTest` (written first — 4 of its 5 cases fail against the previous implementation):

- multi-byte character straddling the buffer boundary → byte-identical copy
- the same case asserted at text level, so a regression reads as "replacement character introduced" rather than a byte diff
- source returning short reads → byte-identical copy
- multi-byte character split across concatenated streams (`SequenceInputStream`, as report results are assembled) → byte-identical copy
- empty stream

Full suite: 311 tests, 0 failures, including the Cucumber e2e scenarios that write and fetch report calculation data against real containers.

## Deploying this does not repair existing data

The helper sits on the **write** path too (`CouchDbFileStorage.storeFile`), so stored `data.json` attachments already contain `EF BF BD` on disk. Affected report calculations must be re-run after this is deployed. They can be found by fetching `data.json` directly from CouchDb, bypassing the API, and searching for `EF BF BD`.

Worth planning for: repairing an attachment changes its digest, and `ReportCalculationChangeUseCase` notifies webhook subscribers on digest change — so recalculating will fire webhook notifications for every affected report.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved stream handling to preserve raw binary data during copying.
  * Prevented corruption of multi-byte text and data when input is read in partial chunks.
  * Ensured copied output is flushed reliably.

* **Tests**
  * Added coverage for empty inputs, short reads, concatenated streams, and UTF-8 boundary cases.

### CODE REVIEW

**Comment by sleidig** [comment]
@coderabbitai review

**Comment by sleidig** [comment]
@coderabbitai review

File: application/aam-backend-service/src/test/kotlin/com/aamdigital/aambackendservice/common/stream/StreamHelperTest.kt [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+import java.io.ByteArrayInputStream
+import java.io.ByteArrayOutputStream
+import java.io.InputStream
+import java.io.SequenceInputStream
+import java.util.Collections
+
+class StreamHelperTest {
```
[the remark this answers is not in the document]
Comment (sleidig) [reply application/aam-backend-service/src/test/kotlin/com/aamdigital/aambackendservice/common/stream/StreamHelperTest.kt:11]: not needed, because the class doesn't mock anything in itself

**Comment by sleidig** [comment]
Tested on affected live system :heavy_check_mark:
